File: [0790][Medium][Domino_and_Tromino_Tiling]/Domino_and_Tromino_Tiling_2.java
// leetcode time     cost : 1 ms
// leetcode memory   cost : 36.4 MB 
// Time  Complexity: O(logN)
// Space Complexity: O(logN)
// solution 2, matrix expo
class Solution {
    int MOD = 1_000_000_007;

    public int numTilings(int N) {
        int[][] T = new int[][]{{1,0,0,1},{1,0,1,0},{1,1,0,0},{1,1,1,0}};
        return matrixExpo(T, N)[0][0];
    }

    public int[][] matrixMult(int[][] A, int[][] B) {
        int[][] ans = new int[A.length][A.length];
        for (int i = 0; i < A.length; i++)
            for (int j = 0; j < B[0].length; j++) {
                long entry = 0;
                for (int k = 0; k < B.length; k++)
                    entry += (long) A[i][k] * (long) B[k][j] % MOD;
                ans[i][j] = (int) (entry % MOD);
            }

        return ans;
    }

    public int[][] matrixExpo(int[][] A, int pow) {
        int[][] ans = new int[A.length][A.length];
        for (int i = 0; i < A.length; i++) ans[i][i] = 1;
        if (pow == 0) return ans;
        if (pow == 1) return A;
        if (pow % 2 == 1) return matrixMult(matrixExpo(A, pow-1), A);
        int[][] B = matrixExpo(A, pow / 2);
        return matrixMult(B, B);
    }
}

File: [0790][Medium][Domino_and_Tromino_Tiling]/Domino_and_Tromino_Tiling_2.py
# leetcode time     cost : 48 ms
# leetcode memory   cost : 13.5 MB 
# Time  Complexity: O(N)
# Space Complexity: O(1)
# solution 1, DP
class Solution(object):
    def numTilings(self, N):
        MOD = 10**9 + 7
        # dp[i][0] both rows are covered for column i,
        # dp[i][1] up row is covered  for column i,
        # dp[i][2] up row is covered  for column i, dp[i][1] == dp[i][2]
        dp = [[0]*3 for _ in range(N+1)]
        # init dp, 
        dp[0][0] = dp[1][0] = 1
        for i in range(2,N+1):
            dp[i][0] = (dp[i-1][0] + dp[i-2][0] + dp[i-1][1] +dp[i-1][2]) % MOD
            dp[i][1] = (dp[i-2][0] + dp[i-1][2]) % MOD
            dp[i][2] = (dp[i-2][0] + dp[i-1][1]) % MOD

        return dp[N][0]

def main():
    n = 3      #expect is 5
    obj = Solution()
    res = obj.numTilings(n)
    print("return value is ",res);
    
if __name__ =='__main__':
    main()   